Dockerfile

```diff
@@ -88,18 +93,35 @@ ENV PIP_SRC=/deps/src/
 ENV PYTHONUSERBASE=/deps
 ENV PATH $PYTHONUSERBASE/bin:$PATH
 ENV NPM_CONFIG_PREFIX=/deps/
-RUN ln -s ${HOME}/package.json /deps/package.json \
+ENV NPM_CACHE_DIR=/deps/cache/npm
+ENV NPM_DEBUG=true
+
+RUN \
+    # Files needed to run the make command
+    --mount=type=bind,source=Makefile,target=${HOME}/Makefile \
+    --mount=type=bind,source=Makefile-docker,target=${HOME}/Makefile-docker \
+    # Files required to install pip dependencies
+    --mount=type=bind,source=setup.py,target=${HOME}/setup.py \
+    --mount=type=bind,source=./requirements,target=${HOME}/requirements \
+    # Files required to install npm dependencies
+    --mount=type=bind,source=package.json,target=${HOME}/package.json \
+    --mount=type=bind,source=package-lock.json,target=${HOME}/package-lock.json \
+    # Mounts for caching pip/npm dependencies
+    --mount=type=cache,target=${PIP_CACHE_DIR},uid=${OLYMPIA_UID},gid=${OLYMPIA_GID} \
+    --mount=type=cache,target=${NPM_CACHE_DIR},uid=${OLYMPIA_UID},gid=${OLYMPIA_GID} \
+    # Command to install pip/npm dependencies
+    ln -s ${HOME}/package.json /deps/package.json \
     && ln -s ${HOME}/package-lock.json /deps/package-lock.json \
     && make update_deps_prod
 
+# Only copy our source files after we have installed all dependencies
```

Comment:
What difference does that make ?

Reply:
The difference here is subtle and not totally visible yet but the further up in the build chain we copy "." the more likely it is that we invalidate steps which don't actually depend on 100% of files.

By moving this instruction to the final and bottom most step we can allow docker to more efficiently cache individual layers which means we are much more likely to get a cache hit.

For example, in the updated version of this pr you can see that the locale compilation step is in a separate stage with only the files it needs copied, this means the layer will only rebuild if you change files in that specific folder.

We will get the absolute most out of this change once we are able to extract the `update_assets` command from the final stage as this takes about 30-45 seconds and because it runs after copying will re-run no matter what files change. But the work to do that is way too big for this PR so we can do that later.